lib/ecto/changeset.ex

```diff
@@ -1136,6 +1137,35 @@ defmodule Ecto.Changeset do
     end)
   end
 
+  @doc """
+  Applies the changeset action only if the changes are valid.
+
+  If the changes are valid, all changes are applied to the changeset data.
+  If the changes are invalid, no changes are applied, and an error tuple
+  is returned with the changeset containing the action that was attempted
+  to be applied.
+
+  The action may be one of #{Enum.map_join(@actions, ", ", &"`#{inspect &1}`")}.
+
+  ## Examples
+
+      iex> {:ok, data} = apply_action(changeset, :update)
+
+      iex> {:error, changeset} = apply_action(changeset, :update)
+      %Ecto.Changeset{action: :update}
+  """
+  @spec apply_action(t, action) :: {:ok, data} | {:error, Ecto.Schema.t}
```

Comment:
I think the type spec should be:

```elixir
 :: {:ok, data} | {:error, t}
```

or (similar to apply_changes)

```elixir
 :: {:ok, Ecto.Schema.t | data} | {:error, t}
```

PS awesome addition!